### ISSUE TITLE
Constant folding doesn't fold weights that are using weight normalization 

### ISSUE THREAD

**Issue by trevor-m** [tensorflow/tensorflow#24083]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 16.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: No
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): 1.12.0
- Python version: 3.5.2
- Bazel version (if compiling from source): 0.15.0?
- GCC/Compiler version (if compiling from source):  5.4.0
- CUDA/cuDNN version: CUDA 10.0.130, CUDNN 7.4.1.5
- GPU model and memory: TITAN V 10956 MB

**Describe the current behavior**
I have a 1D convolution layer whose weights are using weight normalization (Salimans & Kingma, 2016). The weights to the convolution are processed using the following equation: `w = g * v/2-norm(v)`. After running the constfold optimizer, these operations on the weights are still present, even though they should be folded.

**Describe the expected behavior**
I would expect the normalization ops to be folding into the weights, so that only one Const/read node remains.

**Code to reproduce the issue**
```python
import tensorflow as tf
import math
from tensorflow.core.protobuf import meta_graph_pb2
from tensorflow.core.protobuf import rewriter_config_pb2
from tensorflow.python.grappler import tf_optimizer

def conv_graph(x, output_name='output', kernel_width=3, in_dim=1024, out_dim=1024):
  """Applies convolution with gated linear units on x.
  https://github.com/NVIDIA/OpenSeq2Seq/blob/master/open_seq2seq/parts/convs2s/conv_wn_layer.py
    Args:
      x: A float32 tensor with shape [batch_size, length, in_dim]
    Returns:
      float32 tensor with shape [batch_size, length, out_dim].
  """
  # Define Variables
  conv_out_size = 2 * out_dim
  V_std = math.sqrt(4.0 * 0.8 / (kernel_width * in_dim))
  V = tf.get_variable('V', shape=[kernel_width, in_dim, conv_out_size],
                      initializer=tf.random_normal_initializer(mean=0, stddev=V_std),
                      trainable=True)
  V_norm = tf.norm(V.initialized_value(), axis=[0, 1])
  g = tf.get_variable('g', initializer=V_norm, trainable=True)
  W = tf.reshape(g, [1, 1, conv_out_size]) * tf.nn.l2_normalize(V, [0, 1])

  output = tf.nn.conv1d(value=x, filters=W, stride=1, padding="VALID")
  output = tf.identity(output, name=output_name)
  return output

def apply_constfold(frozen_graph, output_nodes):
  graph = tf.Graph()
  with graph.as_default():
    tf.import_graph_def(frozen_graph, name="")
  grappler_meta_graph_def = tf.train.export_meta_graph(graph_def=graph.as_graph_def(add_shapes=True), graph=graph)

  _to_bytes = lambda s: s.encode("utf-8", errors="surrogateescape")
  output_collection = meta_graph_pb2.CollectionDef()
  output_list = output_collection.node_list.value
  for i in output_nodes:
    if isinstance(i, tf.Tensor):
      output_list.append(_to_bytes(i.name))
    else:
      output_list.append(_to_bytes(i))
  # TODO(laigd): use another key as the outputs are really not train_op.
  grappler_meta_graph_def.collection_def["train_op"].CopyFrom(output_collection)
  rewriter_config = rewriter_config_pb2.RewriterConfig()
  rewriter_config.optimizers.extend(["constfold"])

  session_config_with_trt = tf.ConfigProto()
  session_config_with_trt.graph_options.rewrite_options.CopyFrom(
      rewriter_config)
  frozen_graph = tf_optimizer.OptimizeGraph(session_config_with_trt, grappler_meta_graph_def, graph_id=b"tf_graph")
  return frozen_graph

if __name__ == '__main__':
  with tf.Graph().as_default():
    # Create graph
    x = tf.placeholder(dtype=tf.float32, shape=(None, None, 1024), name='input')
    y = conv_graph(x)
    # Initialize
    with tf.Session() as sess:
      sess.run(tf.global_variables_initializer())
      # Freeze graph
      frozen_graph = tf.graph_util.convert_variables_to_constants(
          sess,
          sess.graph_def,
          output_node_names=['output'])

  print('Nodes before:')
  [print(n.name, n.op) for n in frozen_graph.node]

  # const folding
  frozen_graph = apply_constfold(frozen_graph, output_nodes=['output'])

  print('----------------------------------------')
  print('Nodes after:')
  [print(n.name, n.op) for n in frozen_graph.node]
```
**Output of script**
```
Nodes before:
input
V
V/read
g
g/read
Reshape/shape
Reshape
l2_normalize/Square
l2_normalize/Sum/reduction_indices
l2_normalize/Sum
l2_normalize/Maximum/y
l2_normalize/Maximum
l2_normalize/Rsqrt
l2_normalize
mul
conv1d/ExpandDims/dim
conv1d/ExpandDims
conv1d/ExpandDims_1/dim
conv1d/ExpandDims_1
conv1d/Conv2D
conv1d/Squeeze
output
----------------------------------------
Nodes after:
input
V
Reshape
l2_normalize/Sum/reduction_indices
l2_normalize/Maximum/y
conv1d/ExpandDims/dim
conv1d/ExpandDims_1/dim
V/read
conv1d/ExpandDims
l2_normalize/Square
l2_normalize/Sum
l2_normalize/Maximum
l2_normalize/Rsqrt
l2_normalize
mul
conv1d/ExpandDims_1
conv1d/Conv2D
conv1d/Squeeze
output
```

Edit: updated repro for 1.13

**Comment by trevor-m** [comment]
@rmlarsen Have you had any chance to look at this?

**Comment by trevor-m** [comment]
Running with TF_CPP_MIN_VLOG_LEVEL=1, I get the following output from constant_folding.cc. Looks like the size would be 'too large'. The resulting weights would only be [1, 3, 1024, 2048] (around 20mb). TF-TRT heavily relies on constant folding, is there a way we can override this size limit?

```
2018-12-13 20:49:37.173127: I tensorflow/core/grappler/optimizers/constant_folding.cc:1236] Failed to fold node name: "V/read"
op: "Identity"
input: "V"
attr {
  key: "T"
  value {
    type: DT_FLOAT
  }
}

Error message: Invalid argument: Can't fold ConstantFolding/V/read-folded, its size would be too large
```
